File: can_place_flowers.rs
//! <https://leetcode.com/problems/can-place-flowers/>
pub fn can_place_flowers(flowerbed: Vec<i32>, n: i32) -> bool {
    let mut zeros = 0;
    let mut count = 0;
    let mut prev = 0;
    for flower in flowerbed {
        if flower == 0 {
            zeros += 1;
        } else {
            count += (zeros - if prev == 1 { 1 } else { 0 }) / 2;
            prev = flower;
            zeros = 0;
        }
    }
    if zeros != 0 {
        count += (zeros + if prev == 0 { 1 } else { 0 }) / 2;
    }
    count >= n
}

fn main() {
    assert!(can_place_flowers(vec![1, 0, 0, 0, 1], 1));
    assert!(can_place_flowers(vec![1, 0, 0, 0, 0, 0, 0, 0, 1], 1));
    assert!(can_place_flowers(vec![0, 0, 0, 0, 1], 2));
    assert!(can_place_flowers(vec![0, 0, 0, 0, 0], 3));
    assert!(can_place_flowers(vec![0, 1, 0, 0, 0], 1));
    assert!(!can_place_flowers(vec![0, 1, 0, 0, 0], 2));
    assert!(!can_place_flowers(vec![1, 0, 0, 0, 0, 1], 2));
    assert!(!can_place_flowers(vec![1, 0, 0, 0, 1], 2));
    assert!(!can_place_flowers(vec![1], 1));
    assert!(can_place_flowers(vec![0], 1));
}

File: python/can_place_flowers.py
"""
https://leetcode.com/problems/can-place-flowers/
"""

from typing import List


def can_place_flowers(flowerbeds: List[int], n: int) -> bool:
    """
    :type flowerbed: List[int]
    :type n: int
    :rtype: bool
    """
    if flowerbeds[0] == 0:
        flowerbeds = [1, 0] + flowerbeds
    if flowerbeds[-1] == 0:
        flowerbeds += [0, 1]
    num_zeros = 0
    for i in range(1, len(flowerbeds)):
        if flowerbeds[i] == 0:
            num_zeros += 1
        else:
            n -= (num_zeros - 1) / 2
            num_zeros = 0
    return n <= 0


if __name__ == "__main__":
    assert can_place_flowers([1, 0, 0, 0, 1], 1)
    assert not can_place_flowers([1, 0, 0, 0, 1], 2)
